torchtune/models/clip/_text_encoder.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import nn, Tensor
+
+from torchtune.modules import (
+    FeedForward,
+    MultiHeadAttention,
+    TransformerSelfAttentionLayer,
+)
+from torchtune.modules.activations import QuickGELU
+
+
+class CLIPTextEncoder(nn.Module):
+    """
+    Text encoder for CLIP.
+
+    Args:
+        vocab_size (int): size of the vocabulary, default 49408
+        max_seq_len (int): context size, default 77
+        embed_dim (int): embedding/model dimension size, default 768
+        num_heads (int): number of attention heads, default 12
+        num_layers (int): number of transformer layers, default 12
+        norm_eps (float): small value added to denominator for numerical stability, default 1e-5
+    """
+
+    def __init__(
+        self,
+        vocab_size: int = 49408,
+        max_seq_len: int = 77,
+        embed_dim: int = 768,
+        num_heads: int = 12,
+        num_layers: int = 12,
+        norm_eps: float = 1e-5,
+    ):
+        super().__init__()
+        self.max_seq_len = max_seq_len
+
+        self.token_embedding = nn.Embedding(vocab_size, embed_dim)
+        self.position_embedding = nn.Parameter(torch.empty(max_seq_len, embed_dim))
+
+        self.encoder = nn.Sequential(
+            *[
+                TransformerSelfAttentionLayer(
+                    attn=MultiHeadAttention(
+                        embed_dim=embed_dim,
+                        num_heads=num_heads,
+                        num_kv_heads=num_heads,
+                        head_dim=embed_dim // num_heads,
+                        q_proj=nn.Linear(embed_dim, embed_dim),
+                        k_proj=nn.Linear(embed_dim, embed_dim),
+                        v_proj=nn.Linear(embed_dim, embed_dim),
+                        output_proj=nn.Linear(embed_dim, embed_dim),
+                    ),
+                    mlp=FeedForward(
+                        gate_proj=nn.Linear(embed_dim, embed_dim * 4),
+                        down_proj=nn.Linear(embed_dim * 4, embed_dim),
+                        activation=QuickGELU(),
+                    ),
+                    sa_norm=nn.LayerNorm(embed_dim, eps=norm_eps),
+                    mlp_norm=nn.LayerNorm(embed_dim, eps=norm_eps),
+                )
+                for _ in range(num_layers)
+            ]
+        )
+
+        self.final_norm = nn.LayerNorm(embed_dim, eps=norm_eps)
+
+    def forward(self, tokens: Tensor) -> Tensor:
```

Comment:
Should this forward call support all the other things in our `TransformerDecoder` (mask, encoder_input, encoder_mask, input_pos)? Or should we implement these other things on an as-needed basis? For diffusion, this should be sufficient.

Likewise, should I add support for a KV-cache?

Reply:
This doesn't need the same forward signature as TransformerDecoder as it won't be used the same way, but I think it should have a supset of that, maybe "tokens, mask, and input_pos". input_pos could be used instead of argmax for the forward. Or you can just add a todo around that for now and move forward as is.